### PR TITLE
Upload MIP-RNA to all DNA cases in collaborator group matching by subject_ID

### PR DESCRIPTION
**Pull request by peterpru**
## Description
We want to upload all mip-RNA cases to DNA cases, matching on subject_id and tumour yes/no, to all customers within collaborator groups.

Linked issue: https://github.com/Clinical-Genomics/project-planning/issues/316

### Added

-

### Changed

- Query to upload RNA cases matches a list of customer IDs instead of only the same customer ID.

### Fixed

- Function `get_samples_by_customer_id_list_and_subject_id_and_is_tumour` did not check tumour state, only tumour=True, it now checks tumour state from the input.

### How to prepare for test
We are doing the full test. 
1) On stage cg master branch: A MIP-RNA trio (smoothtuna), which is going to be uploaded to two MIP-DNA trios of the same customer (onkiwi, stablecorgi). 
2) Then we change customer of stablecorgi, which should then only upload to onkiwi. 
3) Install this PR branch, and do the upload again as described in test 2. It should upload to both customers.

- [ ] Ssh to hasta
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install **master** on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t cg -b master -a
    ```
- [ ]  get the associated DNA files to stage for onkiwi `mkdir -p /home/proj/stage/housekeeper-bundles/onkiwi/2021-12-24/` and `cp -r /home/proj/production/housekeeper-bundles/onkiwi/2021-12-24/ /home/proj/stage/housekeeper-bundles/onkiwi/2021-12-24/`
- [ ] do `cg upload -f onkiwi`
- [ ] `mkdir /home/proj/stage/housekeeper-bundles/stablecorgi/2022-12-15/` and `cp -r /home/proj/production/housekeeper-bundles/stablecorgi/2022-12-15/ /home/proj/stage/housekeeper-bundles/stablecorgi/2022-12-15/`
- [ ] do `cg upload -f stablecorgi`
- [ ] get the associated RNA files to stage `mkdir -p /home/proj/stage/housekeeper-bundles/smoothtuna/2022-09-02/` and `cp -r /home/proj/production/housekeeper-bundles/smoothtuna/2022-09-02/ /home/proj/stage/housekeeper-bundles/smoothtuna/2022-09-02/ and cp /home/proj/production/housekeeper-bundles/smoothtuna/2022-09-02/smoothtuna_report.pdf /home/proj/stage/housekeeper-bundles/smoothtuna/2022-09-02/`

### How to test
Subject_id C000357840 & C000357838 have only 1 DNA and 1 RNA sample connected. SubjectID C000350464 has 5 samples connected, of which 3 RNA and 2 DNA. Removed subject ID C000350464 from DNA sample ACC10362A7, to make it 1 DNA sample connected. The extra RNA samples should be not be relevant for the upload.

#### Test 1, upload as normal on master.
- [ ] Do `cg upload rna-to-scout smoothtuna`

#### Test 2, upload on master after changing customers in statusDB stage for case stablecorgi.
- [ ] Change customer in statusDB stage for stablecorgi from 3 to 4.
- [ ] Do `cg upload rna-to-scout smoothtuna`

#### Test 3, install this PR branch, and do the upload again.
- [ ] Install this branch:
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t cg -b upload-mip-rna-to-collaborators -a
    ```
- [ ] Do `cg upload rna-to-scout smoothtuna`

#### Test 1, upload as normal on master.
- [x] Check that all relevant files are uploaded to 2 DNA cases.
- [x] Take a screenshot and attach or copy/paste the output. 
image

#### Test 2, only 1 DNA case connected for the relevant customer.
- [x] Check that all relevant files are uploaded to onkiwi only.
- [x] Take a screenshot and attach or copy/paste the output. 
image

#### Test 3, RNA case is uploaded to 2 DNA cases, each being from a different customer.
- [x] Check that all relevant files are uploaded to 2 DNA cases.
- [x] Take a screenshot and attach or copy/paste the output. 
image

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

**Comment by peterpru** [comment]
New multiQC report:
image

**Comment by peterpru** [comment]
We should have a fail-safe, where it first checks that there is a DNA case available for the RNA customer, so that it not only uploads to DNA cases from other customers. (Sofia had a case like this today in prod). Otherwise The RNA customer will not get access to their own data.

**Review by seallard — APPROVED** [review]
👍

**Review by diitaz93 — APPROVED** [review]
Nice catch! It works, but I made some suggestions anyway 😄

File: cg/meta/upload/scout/uploadscoutapi.py [hunk 1]
```diff
@@ -458,7 +461,7 @@ def _add_dna_cases_to_dna_sample(
             if (
                 case_object.data_analysis
                 in [Pipeline.MIP_DNA, Pipeline.BALSAMIC, Pipeline.BALSAMIC_UMI]
-                and case_object.customer_id == rna_sample.customer_id
+                and case_object.customer in rna_sample.customer.collaborators
```
[the remark this answers is not in the document]
Comment (peterpru) [reply cg/meta/upload/scout/uploadscoutapi.py:464]: Nice, added this to the new PR as well. https://github.com/Clinical-Genomics/cg/pull/1988

File: cg/store/api/find_business_data.py [hunk 2]
```diff
@@ -666,19 +666,27 @@ def get_samples_by_customer_subject_id_and_is_tumour(
...
+                SampleFilter.FILTER_BY_CUSTOMER_ENTRY_IDS,
+                SampleFilter.FILTER_BY_SUBJECT_ID,
+                SampleFilter.FILTER_IS_TUMOUR,
+            ]
+        else:
+            filter_functions = [
+                SampleFilter.FILTER_BY_CUSTOMER_ENTRY_IDS,
+                SampleFilter.FILTER_BY_SUBJECT_ID,
+                SampleFilter.FILTER_IS_NOT_TUMOUR,
+            ]
```
[the remark this answers is not in the document]
Comment (peterpru) [reply cg/store/api/find_business_data.py:684]: Nice, I have made a new PR: https://github.com/Clinical-Genomics/cg/pull/1988

**Comment by peterpru** [comment]
Successfully deployed in stage:
image

**Comment by peterpru** [comment]
Deployed in production:
image

**Pull request merged**